### ISSUE TITLE
Incorrect base class detection

### ISSUE THREAD

**Issue by adamjstewart** [sphinx-doc/sphinx#9607]
### Describe the bug

I have a chain of subclasses, like A -> B -> C -> D. When I document class D, it lists the base class as A instead of C.

### How to Reproduce

```
$ git clone https://github.com/microsoft/torchgeo.git
$ cd torchgeo/docs
$ pip install -r requirements.txt
$ make html
$ # open _build/html/api/datasets.html and see that the base class is torch.utils.data.Dataset instead of RasterDataset
```

### Expected behavior

I would expect the base class to be the direct super class.

### Your project

https://github.com/microsoft/torchgeo

### OS

macOS 10.15.7, Linux (whatever RtD uses)

### Python version

3.8.11

### Sphinx version

4.0.1, 4.0.2

### Sphinx extensions

sphinx.ext.autodoc

### Additional context

I don't think this is the same as #9504 because I can reproduce it with 4.0.1 too.

Could be related to #9395.

For a class like [Landsat](https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/landsat.py#L14), even though the super class is `RasterDataset`, the base class is listed as `torch.utils.data.Dataset`.

@calebrob6

**Comment by tk0miya** [comment]
It seems `Landsat.__orig_bases__` is incorrect. It should be `(RasterDataset, abc.ABC)`. But it returns `(Dataset, abc.ABC)` instead. It must be a bug of Python interpreter.

```
$ python
Python 3.8.12 (default, Sep  3 2021, 02:24:44)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from torchgeo.datasets import Landsat
>>> Landsat.__mro__
(<class 'torchgeo.datasets.Landsat'>, <class 'torchgeo.datasets.RasterDataset'>, <class 'torchgeo.datasets.GeoDataset'>, <class 'torch.utils.data.Dataset'>, <class 'typing.Generic'>, <class 'abc.ABC'>, <class 'object'>)
>>> Landsat.__orig_bases__
(torch.utils.data.Dataset[typing.Dict[str, typing.Any]], <class 'abc.ABC'>)
```

I got the same result with 3.9.7.
```
$ python
Python 3.9.7 (default, Sep  3 2021, 02:02:37)
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from torchgeo.datasets import Landsat
>>> Landsat.__mro__
(<class 'torchgeo.datasets.Landsat'>, <class 'torchgeo.datasets.RasterDataset'>, <class 'torchgeo.datasets.GeoDataset'>, <class 'torch.utils.data.Dataset'>, <class 'typing.Generic'>, <class 'abc.ABC'>, <class 'object'>)
>>> Landsat.__orig_bases__
(torch.utils.data.Dataset[typing.Dict[str, typing.Any]], <class 'abc.ABC'>)
```

Note: This is a Dockefile for reproduce.
```
FROM python:3.8-slim

RUN apt update; apt install -y build-essential curl git make unzip vim
RUN git clone https://github.com/microsoft/torchgeo.git
WORKDIR /torchgeo/docs
RUN pip install -r requirements.txt
RUN make html
```

**Comment by tk0miya** [comment]
I found a workaround for Python's behavior (refs: #9654). It will be merged and fixed in 4.3.0.
